.github/workflows/on_push_clang_format.yml

```diff
@@ -0,0 +1,11 @@
+name: Clang Format Checker
+on: [push]
+jobs:
+  clang-format-checking:
+    runs-on: ubuntu-latest
+    steps:
+      - uses: actions/checkout@v2
+      - uses: RafikFarhad/clang-format-github-action@v2.0.0
```

Comment:
The truth is that I just used the first Github Action that I encountered. I'll try this one out :+1:

Reply:
I have been playing a bit around with `clang-format-lint-action` and it works quite well! It runs faster than the other one and as you said we can think about integrating an automatic action to automatically apply clang-format before merging a PR. 

However, I would wait a bit for taking the step of introducing automatic commits related to the clang-format ... it scares me a bit what could happen without having some weeks/months of experience with the new clang-format settings we are defining here. 

What I can definitely do is write some documentation about the implications of enforcing clang-format now in the project in the file `CONTRIBUTING.md`.

Reply:
@piponazo You are probably right that it's better to gain some experience with the action before we let it automatically commit to master. 

How would you propose the workflow should look?       
Let's say a user opens a PR and the pipeline fails, what would you suggest to them they do?     
What if they don't have `clang-format`?     
Or like in my case my system comes with 
```sh
❯ clang-format --version
clang-format version 13.0.1
```
That means for me there will be cases where I won't be able to fix the pipeline without installing a new `clang-format` version or trying to figure out the correct diff to apply manually from the failing ci log.   

All of that sounds like a bit too much friction to me.   

What if additionally to the check that you implemented, we add one more action which is triggered by a specific comment. 
E.g. if a comment includes `ExivBot plz format` the action will be triggered and applies the correct formatting to the PR.

Reply:
I was thinking about this yesterday when I went to bed (I think I have a problem 🛌 ).

If a contributor creates a PR and the CI pipeline fails (this might happen not only to new contributors but also to experienced ones) we will have a a CI job indicating exactly what is the problem. At that point we can point them to the `CONTRIBUTING.md` file which has some notes about applying clang-format before summiting code.

Regarding the clang-format versions. I tried yesterday and clang-format-12 & clang-format-13 are producing the same results with the `.clang-format` file we have here. Of course, if they do not have clang-format installed, they will need to go to the LLVM download page and download the binaries for their platform: https://releases.llvm.org/download.html. I agree that this might create some friction for people who never used clang-format before. But hey if they managed to install git, compilers, cmake, conan, etc ... they should manage to install one tool more 😋 .

The main problems I see with the automatic approach are the following:
- A contributor creates a PR and we will not have a clang-format check at that point (the automatic formatting will happen after the PR is approved and merged). Since we have no formatting checks when the PR is created, we will see PRs that might contain some crazy formatting , reviewers might start to discuss again about formatting style, or it is simply more difficult to review the code because we see different formatting styles in each PR. For me that is one of the most important point about brining a unified and enforced clang-format: to easy the coding reviews.
- Let's say that we can "live" with the previous point. After the PR is approved and we click the merge button, the automatic commit happens. At this point at least 2 things might go wrong:
  - The automatic commit produce some changes that break the compilation. This barely happens but the probability is not 0%. It happens twice to me while working on this branch: Once due to the order of headers, and other time with some problem with preprocessor code and comments. 
  - The 2 problem I see is that we might not like the automatic formatting applied to the code and we have no chance to adapt the code in the PR (the PR will be already merged!). By forcing to have the clang-format job passing before merging, we have the chance to still make some suggestions about formatting during the coding review. For example: adding a comma after the last element of an array to have a nicer style after clang-format is applied. 
- We can end up having too many commits just applying clang-format and when you try to use `git blame` you end up seeing that many areas of the code were finally touched by a "ROBOT". I think that this is bad and should be avoided as much as possible. We should also recommend contributors to apply clang-format before committing. Or if they forget, it is always possible to do an interactive rebase and squash some commits together.

Reply:
Behavior between versions can definitely change, we've experienced this at work.            
Thus, to be on the safe side one has to have the correct version, which is something I don't like.  I personally don't want to have to download a specific version of clang-format for exiv2 and then make my editor somehow pick that version over my default one. 

I think the contributing guidelines can still say to apply `clang-format` if it is available on a user's system. That get's a good starting point for a review  

Otherwise, I agree with some of your points about the automatic commit, but my last comment was about a manually triggered action on a PR. 

This could be applied before a review if the original formatting is terrible, or at the end before merging. 
But it would still enable a review of the formatting before merging and it would trigger the CI so we know if it compiles. 
And the problem of having these commits in the history, yes that is a bit annoying but in most cases we can probably just squash & merge and thus won't have those commits.  

A PR that is big enough to have many commits that shouldn't be squashed probably comes from a user who format and/or rebase themselves.

Reply:
Right, I think the approach about having a comment to trigger the clang-format in the PR might be a good solution to achieve the best of both sides. 

I'll switch this PR back to draft state and play a bit with that 😉